rir/src/compiler/analysis/verifier.cpp

```diff
@@ -309,6 +309,32 @@ class TheVerifier {
                 ok = false;
             }
         });
+
+        verifyEffects(i);
+    }
+
+    void verifyEffects(Instruction* i) {
+        Effects eff = i->effects;
+        Effects effAndEnv = eff;
+        effAndEnv.set(Effect::LeakArg);
+        effAndEnv.set(Effect::ReadsEnv);
+        effAndEnv.set(Effect::WritesEnv);
+        effAndEnv.set(Effect::LeaksEnv);
+
+        auto fail = [&](auto msg) {
+            std::cerr << "Error in effects '" << eff << "', in:\n";
+            i->print(std::cerr);
+            std::cerr << "\n" << msg << "\n";
+            ok = false;
+        };
+
+        if (eff.contains(Effect::ExecuteCode) && effAndEnv != Effects::Any()) {
+            fail("ExecuteCode implies all other effects except env-based");
+        }
+
+        if (eff.contains(Effect::LeaksEnv) && !eff.contains(Effect::LeakArg)) {
+            fail("LeaksEnv implies LeakArg");
+        }
```

Comment:
honestly I don't agree with either of the verifications.

e.g. we could have a call, that we know the target. so the call instruction would have as effects executeCode + whatever the callee does. but that does not have to be Any.

then I am not sure leakArg includes the environment. I guess we are not heavily using it yet. but I can see that we might want leakArg to mean leak some argument other than the env.